pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,205 @@
+		backoffRecord := b.(*backoffRecord)
+		backoffRecord.failuresAfterLastSuccess = record.failuresAfterLastSuccess
+		backoffRecord.lastFailureTime = record.lastFailureTime
+	}
+
+	return nil
+}
+
+func (s *backoffStore) removeBackoffRecord(jobId string) error {
+	b, ok, err := s.store.GetByKey(jobId)
+	if err != nil {
+		return err
+	}
+
+	if ok {
+		err = s.store.Delete(b)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+
+}
+
+func newBackoffRecordStore() *backoffStore {
+	return &backoffStore{
+		store: cache.NewStore(backoffRecordKeyFunc),
+	}
+}
+
+var backoffRecordKeyFunc = func(obj interface{}) (string, error) {
+	if u, ok := obj.(*backoffRecord); ok {
+		return u.key, nil
+	}
+	return "", fmt.Errorf("could not find key for obj %#v", obj)
+}
+
+func (backoffRecordStore *backoffStore) newBackoffRecord(clock clock.WithTicker, key string, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffRecord {
+	now := clock.Now()
+	var backoff *backoffRecord
+
+	if b, exists, _ := backoffRecordStore.store.GetByKey(key); exists {
+		old := b.(*backoffRecord)
+		backoff = &backoffRecord{
+			key:                      old.key,
+			failuresAfterLastSuccess: old.failuresAfterLastSuccess,
+			lastFailureTime:          old.lastFailureTime,
+		}
+	} else {
+		backoff = &backoffRecord{
+			key:                      key,
+			failuresAfterLastSuccess: 0,
+			lastFailureTime:          nil,
+		}
+	}
+
+	sortByFinishedTime(newSucceededPods, now)
+	sortByFinishedTime(newFailedPods, now)
+
+	if len(newSucceededPods) == 0 {
+		if len(newFailedPods) == 0 {
+			return *backoff
+		}
+
+		backoff.failuresAfterLastSuccess = backoff.failuresAfterLastSuccess + int32(len(newFailedPods))
+		lastFailureTime := getFinishedTime(newFailedPods[len(newFailedPods)-1], now)
+		backoff.lastFailureTime = &lastFailureTime
+		return *backoff
+
+	} else {
+		if len(newFailedPods) == 0 {
+			backoff.failuresAfterLastSuccess = 0
+			backoff.lastFailureTime = nil
+			return *backoff
+		}
+
+		backoff.failuresAfterLastSuccess = 0
+		backoff.lastFailureTime = nil
+
+		lastSuccessTime := getFinishedTime(newSucceededPods[len(newSucceededPods)-1], now)
+		for i := len(newFailedPods) - 1; i >= 0; i-- {
+			failedTime := getFinishedTime(newFailedPods[i], now)
+			if !failedTime.After(lastSuccessTime) {
+				break
+			}
+			if backoff.lastFailureTime == nil {
+				backoff.lastFailureTime = &failedTime
+			}
+			backoff.failuresAfterLastSuccess += 1
+		}
+
+		return *backoff
+
+	}
+
+}
+
+func sortByFinishedTime(pods []*v1.Pod, currentTime time.Time) {
+	sort.Slice(pods, func(i, j int) bool {
+		p1 := pods[i]
+		p2 := pods[j]
+		p1FinishTime := getFinishedTime(p1, currentTime)
+		p2FinishTime := getFinishedTime(p2, currentTime)
+
+		return p1FinishTime.Before(p2FinishTime)
+	})
+}
+
+func getFinishedTime(p *v1.Pod, currentTime time.Time) time.Time {
+	var finishTime *time.Time
+	for _, containerState := range p.Status.ContainerStatuses {
+		if containerState.State.Terminated == nil {
```

Comment:
I guess it could be considered a bug in Kubelet, but I don't think it is being worked on, so we should handle it at the controller side.

Reply:
Please add a unit test for that

Reply:
I think it is just zero for time, time.Time{}, but you may double check

Reply:
right, it's not a pointer, so it looks like it's just epoch :thinking: 

If so, it might be problematic for calculating the backoff (as in theory it would have passed).

Reply:
but then unit tests need to change, because they are actually setting epoch

Reply:
I think they should change to mimic the common scenarios when FinishedAt is non-zero. The edge case should have a dedicated test case IMO.

Reply:
Done